### PR TITLE
Fixes for some h2spec issues

### PR DESCRIPTION
**Pull request by djphoenix**
Issue: #22 
Fixed:
```
5.1 Stream States
    × idle: Sends a DATA frame
    × idle: Sends a RST_STREAM frame
    × idle: Sends a WINDOW_UPDATE frame
5.1.1 Stream Identifiers
    × Sends even-numbered stream identifier
5.5 Extending HTTP/2
    × Sends an unknown extension frame in the middle of a header block
6.1 DATA
    × Sends a DATA frame with 0x0 stream identifier
6.4 RST_STREAM
    × Sends a RST_STREAM frame on a idle stream
6.5.2 Defined SETTINGS Parameters
    × SETTINGS_ENABLE_PUSH (0x2): Sends the value other than 0 or 1
    × SETTINGS_INITIAL_WINDOW_SIZE (0x4): Sends the value above the maximum flow control window size
    × SETTINGS_MAX_FRAME_SIZE (0x5): Sends the value below the initial value
    × SETTINGS_MAX_FRAME_SIZE (0x5): Sends the value above the maximum allowed frame size
8.2. Server Push
    × Sends a PUSH_PROMISE frame
```

### CODE REVIEW

File: lib/spdy-transport/protocol/http2/parser.js [hunk 1]
```diff
@@ -153,8 +166,13 @@ Parser.prototype.onFrameBody = function onFrameBody(header, buffer, callback) {
     this.onPriorityFrame(header, buffer, callback);
   else if (header.type === frameType.X_FORWARDED_FOR)
     this.onXForwardedFrame(header, buffer, callback);
-  else
+  else {
+	if (this._lastHeaderBlock) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Please use 2-space indentation.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 2]
```diff
@@ -153,8 +166,13 @@ Parser.prototype.onFrameBody = function onFrameBody(header, buffer, callback) {
     this.onPriorityFrame(header, buffer, callback);
   else if (header.type === frameType.X_FORWARDED_FOR)
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Please add braces to every `if` condition.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 3]
```diff
@@ -386,6 +419,28 @@ Parser.prototype.onSettingsFrame = function onSettingsFrame(header,
       settings[name] = value;
   }
 
+  if (constants.settingsIndex[constants.settings_server.SETTINGS_ENABLE_PUSH] in settings) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Woooow, this looks a bit long. Any way it could be shrunk with all repeating code removed?

File: lib/spdy-transport/protocol/http2/constants.js [hunk 4]
```diff
@@ -46,7 +47,7 @@ exports.flags = {
   PRIORITY: 0x20
 };
 
-exports.settings = {
+exports.settings_server = {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/constants.js]: Are the `settings` values different for client and server? What's the point in renaming it?

**Comment by indutny** [comment]
Overall I'm really glad that you have decided to work on this! This is fantastic! Thank you.

Let's fix some style issues, and this PR will be ready to land.

File: lib/spdy-transport/protocol/http2/parser.js (hunk 3 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: ACK

File: lib/spdy-transport/protocol/http2/parser.js (hunk 2 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: ACK

File: lib/spdy-transport/protocol/http2/parser.js (hunk 1 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: OK, will fix

File: lib/spdy-transport/protocol/http2/constants.js (hunk 4 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/constants.js]: There is another "settings" export below, that overrides this one. That contains one of parameters with another code, so, I renamed this one. Maybe you suggest another solution?

**Comment by djphoenix** [comment]
CI seems red, is this an test-cases issues or I was broke something?

File: lib/spdy-transport/protocol/http2/constants.js (hunk 4 above)
Comment (indutny) [inline lib/spdy-transport/protocol/http2/constants.js]: Yikes! These are just leftovers from spdy protocol. Please remove them!

File: lib/spdy-transport/protocol/http2/parser.js (hunk 3 above)
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Perhaps you may want to use just a string value from `settingsIndex` directly? Instead of doing two lookups? Please consider not using ` in ` operator too, `settings['...'] !== undefined` should be fine.

File: lib/spdy-transport/protocol/http2/constants.js (hunk 4 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/constants.js]: ACK

**Comment by indutny** [comment]
@djphoenix have you actually tried running `npm test` locally? ;) There seems to be lots of failures. I suspect that this is what will happen with this `even-ids streams` fix. You may want to submit this fix in a separate PR, where we will figure out how to make this change without invalidating existing tests.

File: lib/spdy-transport/protocol/http2/parser.js (hunk 3 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: So, if using strings directly, this may become difficult for changing during refactoring... Are usage of "magic strings" is good idea?

**Comment by djphoenix** [comment]
OK, will work on it...

File: lib/spdy-transport/protocol/http2/parser.js (hunk 3 above)
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: This is absolutely fine. We could as well say that `SETTINGS_ENABLE_PUSH` may be changed, and it will be hard to refactor every place that is using it ;)

The idea behind `settingsIndex` is that it is very convenient to use in the protocol parser, it should not really be used anywhere else. This PR's use case doesn't really look that convenient.

**Comment by djphoenix** [comment]
Previous comments fixed, now working on test failures...

**Comment by djphoenix** [comment]
@indutny any ideas about [/protocol/base/scheduler.js#L128](https://github.com/indutny/spdy-transport/blob/master/lib/spdy-transport/protocol/base/scheduler.js#L128)? It will fix h2spec 5.4.1 (and possible more).

File: lib/spdy-transport/protocol/http2/parser.js [hunk 5]
```diff
@@ -153,6 +166,9 @@ Parser.prototype.onFrameBody = function onFrameBody(header, buffer, callback) {
     this.onPriorityFrame(header, buffer, callback);
   else if (header.type === frameType.X_FORWARDED_FOR)
     this.onXForwardedFrame(header, buffer, callback);
+  else if (this._lastHeaderBlock)
+    return callback(this.error(constants.error.PROTOCOL_ERROR,
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: If `if`'s body does not fit into one line, it should be enclosed in `{` `}` braces.

If you enclose one body of the `if`-chain, all bodies should be enclosed.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 6]
```diff
@@ -386,6 +417,28 @@ Parser.prototype.onSettingsFrame = function onSettingsFrame(header,
...
+  if (settings['initial_window_size'] !== undefined) {
+    if ((settings['initial_window_size'] > constants.MAX_INITIAL_WINDOW_SIZE)
+     || (settings['initial_window_size'] < 0)) {
+      return callback(this.error(constants.error.FLOW_CONTROL_ERROR,
+                                 'SETTINGS_INITIAL_WINDOW_SIZE is OOB'));
+     }
+  }
+  if (settings['max_frame_size']) {
+    if ((settings['max_frame_size'] > constants.ABSOLUTE_MAX_FRAME_SIZE)
+     || (settings['max_frame_size'] < constants.INITIAL_MAX_FRAME_SIZE)) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Does it pass `npm test`? There is a linting phase running after all tests, and it should catch most of the linting errors, including "operator must be at the end of the line".

File: lib/spdy-transport/protocol/http2/parser.js [hunk 7]
```diff
@@ -345,6 +371,11 @@ Parser.prototype.onRSTFrame = function onRSTFrame(header, body, callback) {
     return callback(this.error(constants.error.PROTOCOL_ERROR,
                                'Invalid stream id for RST_STREAM'));
   }
+  
+  if (this._lastHeaderBlock === null && this.pendingHeader === null) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: I don't think that this is correct. It looks like this will just hide the issue, without actually fixing it.

The proper way should be to handle this on connection level, where the information about idle-ness of the stream should be available.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 8]
```diff
@@ -181,6 +197,16 @@ Parser.prototype.unpadData = function unpadData(header, body, callback) {
...
 Parser.prototype.onDataFrame = function onDataFrame(header, body, callback) {
   var isEndStream = (header.flags & constants.flags.END_STREAM) !== 0;
+  
+  if (header.id === 0) {
+    return callback(this.error(constants.error.PROTOCOL_ERROR,
+                               'Received DATA frame with stream=0'));
+  }
+
+  if (this._lastHeaderBlock === null && this.pendingHeader === null) {
+    return callback(this.error(constants.error.PROTOCOL_ERROR,
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Ditto, should be handle on connection layer.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 9]
```diff
@@ -109,6 +111,17 @@ Parser.prototype.onFrameHead = function onFrameHead(buffer, callback) {
     id: buffer.readUInt32BE() & 0x7fffffff
   };
 
+  if (this.isServer && header.id != 0) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: What about clients, should they verify that server-sent PUSH_PROMISE ids are bigger than previous?

File: lib/spdy-transport/protocol/http2/parser.js (hunk 5 above)
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Putting this `if` here will not really help, as far as I could tell, it should be moved to the very top, right after `HEADERS` and `CONTINUATION` handlers. `DATA` handler should be put after it.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 10]
```diff
@@ -386,6 +417,28 @@ Parser.prototype.onSettingsFrame = function onSettingsFrame(header,
...
+  }
+  if (settings['initial_window_size'] !== undefined) {
+    if ((settings['initial_window_size'] > constants.MAX_INITIAL_WINDOW_SIZE)
+     || (settings['initial_window_size'] < 0)) {
+      return callback(this.error(constants.error.FLOW_CONTROL_ERROR,
+                                 'SETTINGS_INITIAL_WINDOW_SIZE is OOB'));
+     }
+  }
+  if (settings['max_frame_size']) {
+    if ((settings['max_frame_size'] > constants.ABSOLUTE_MAX_FRAME_SIZE)
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Could these two `if`s be combined into one?

File: lib/spdy-transport/protocol/http2/parser.js [hunk 11]
```diff
@@ -386,6 +417,28 @@ Parser.prototype.onSettingsFrame = function onSettingsFrame(header,
...
   }
 
+  if (settings['enable_push'] !== undefined) {
+    if ((settings['enable_push'] !== 0)
+     && (settings['enable_push'] !== 1)) {
+      return callback(this.error(constants.error.PROTOCOL_ERROR,
+                                 'SETTINGS_ENABLE_PUSH must be 0 or 1'));
+     }
+  }
+  if (settings['initial_window_size'] !== undefined) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Ditto.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 12]
```diff
@@ -386,6 +417,28 @@ Parser.prototype.onSettingsFrame = function onSettingsFrame(header,
       settings[name] = value;
   }
 
+  if (settings['enable_push'] !== undefined) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Ditto.

Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Another thought, it may be beneficial to introduce `._validateSettings` method, which will enclose all this validators inside of it.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 13]
```diff
@@ -506,6 +562,11 @@ Parser.prototype.onWindowUpdateFrame = function onWindowUpdateFrame(header,
                                'WINDOW_UPDATE length != 4'));
   }
 
+  if (this._lastHeaderBlock === null && this.pendingHeader === null) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Connection-level thing.

**Comment by djphoenix** [comment]
CI seems to be green after merge #24 (SETTINGS validation failed)

**Comment by djphoenix** [comment]
Now fixed:
```
5.1 Stream States
    × idle: Sends a DATA frame
    × idle: Sends a RST_STREAM frame
    × idle: Sends a WINDOW_UPDATE frame
5.1.1 Stream Identifiers
    × Sends even-numbered stream identifier
5.5 Extending HTTP/2
    × Sends an unknown extension frame in the middle of a header block
6.1 DATA
    × Sends a DATA frame with 0x0 stream identifier
6.2. HEADERS
    × Sends a HEADERS frame followed by any frame other than CONTINUATION
6.4 RST_STREAM
    × Sends a RST_STREAM frame on a idle stream
6.5.2 Defined SETTINGS Parameters
    × SETTINGS_ENABLE_PUSH (0x2): Sends the value other than 0 or 1
    × SETTINGS_INITIAL_WINDOW_SIZE (0x4): Sends the value above the maximum flow control window size
    × SETTINGS_MAX_FRAME_SIZE (0x5): Sends the value below the initial value
    × SETTINGS_MAX_FRAME_SIZE (0x5): Sends the value above the maximum allowed frame size
6.10. CONTINUATION
    × Sends a CONTINUATION frame followed by any frame other than CONTINUATION
8.2. Server Push
    × Sends a PUSH_PROMISE frame
```

File: lib/spdy-transport/connection.js [hunk 14]
```diff
@@ -266,6 +266,23 @@ Connection.prototype._handleFrame = function _handleFrame(frame) {
       this.emit('version', 3.1);
     }
     state.window.send.update(frame.delta);
+    if (state.window.send.current > 2147483647) {
```
Comment (indutny) [inline lib/spdy-transport/connection.js]: This number should be in `constants`

File: lib/spdy-transport/connection.js [hunk 15]
```diff
@@ -273,9 +290,15 @@ Connection.prototype._handleFrame = function _handleFrame(frame) {
     // Load created one
     stream = state.stream.map[frame.id];
 
-    // RST or PRIORITY maybe out of sync
-    if (!stream && (frame.type === 'RST' || frame.type === 'PRIORITY'))
+    if (stream === null) {
```
Comment (indutny) [inline lib/spdy-transport/connection.js]: Shouldn't `RST` and `PRIORITY` be ignored on closed streams?

File: lib/spdy-transport/connection.js [hunk 16]
```diff
@@ -284,11 +307,26 @@ Connection.prototype._handleFrame = function _handleFrame(frame) {
...
+      this._goaway({
+        lastId: state.stream.lastId.both,
+        code: 'PROTOCOL_ERROR',
+        send: true
+      });
       return;
     }
   }
 
+  if (!stream && state.isServer && state.version === 4 && frame.id !== 0 &&
```
Comment (indutny) [inline lib/spdy-transport/connection.js]: Please put each condition on a new line (not sure if `jscs` enforces this, but anyway).

File: lib/spdy-transport/connection.js [hunk 17]
```diff
@@ -284,11 +307,26 @@ Connection.prototype._handleFrame = function _handleFrame(frame) {
...
+      (frame.id % 2 === 0 || frame.id < state.stream.lastId.received)) {
```
Comment (indutny) [inline lib/spdy-transport/connection.js]: It should be fine to leave these two conditions on the same line, though, as they are in parens.

File: lib/spdy-transport/connection.js [hunk 18]
```diff
@@ -614,6 +652,7 @@ Connection.prototype._removeStream = function _removeStream(stream) {
 
   state.debug('id=0 remove stream=%d', stream.id);
   delete state.stream.map[stream.id];
+  state.stream.map[stream.id] = null;
```
Comment (indutny) [inline lib/spdy-transport/connection.js]: This is potentially very bad. It will lead to the hashmap that is going to grow with every closed request. Please consider removing this line, and checking against `undefined` everywhere else.

File: lib/spdy-transport/connection.js [hunk 19]
```diff
@@ -629,6 +668,7 @@ Connection.prototype._goaway = function _goaway(params) {
 
   Object.keys(state.stream.map).forEach(function(id) {
     var stream = state.stream.map[id];
+    if (stream === null) return;
```
Comment (indutny) [inline lib/spdy-transport/connection.js]: `undefined`

File: lib/spdy-transport/connection.js [hunk 20]
```diff
@@ -695,6 +735,7 @@ Connection.prototype.destroyStreams = function destroyStreams(err) {
   var state = this._spdyState;
   Object.keys(state.stream.map).forEach(function(id) {
     var stream = state.stream.map[id];
+    if (stream === null) return;
```
Comment (indutny) [inline lib/spdy-transport/connection.js]: `undefined`

File: lib/spdy-transport/connection.js (hunk 15 above)
Comment (indutny) [inline lib/spdy-transport/connection.js]: `undefined`

File: lib/spdy-transport/connection.js (hunk 14 above)
Comment (indutny) [inline lib/spdy-transport/connection.js]: `send.getCurrent()`

File: lib/spdy-transport/protocol/http2/constants.js [hunk 21]
```diff
@@ -17,6 +17,7 @@ exports.INITIAL_MAX_FRAME_SIZE = 16384;
 exports.ABSOLUTE_MAX_FRAME_SIZE = 16777215;
 exports.HEADER_TABLE_SIZE = 4096;
 exports.DEFAULT_MAX_HEADER_LIST_SIZE = 80 * 1024;  // as in http_parser
+exports.MAX_INITIAL_WINDOW_SIZE = 2147483647;
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/constants.js]: This should be duplicated in `spdy/constants.js` if you wish to use it in `connection.js`

File: lib/spdy-transport/protocol/http2/parser.js [hunk 22]
```diff
@@ -18,7 +18,7 @@ function Parser(options) {
   this.waiting = constants.PREFACE_SIZE;
   this.state = 'preface';
   this.pendingHeader = null;
-
+
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Unnecessary change.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 23]
```diff
@@ -154,9 +154,18 @@ Parser.prototype.onFrameBody = function onFrameBody(header, buffer, callback) {
   else if (header.type === frameType.X_FORWARDED_FOR)
     this.onXForwardedFrame(header, buffer, callback);
   else
-    callback(null, { type: 'unknown: ' + header.type });
+    this.onUnknownFrame(header, buffer, callback);
 };
 
+Parser.prototype.onUnknownFrame = function onUnknownFrame(header, buffer, callback) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Erm? What error does this really catch?

I think it should be handling non-`CONTINUATION` frames in-between? Not just unknown frames.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 24]
```diff
@@ -154,9 +154,18 @@ Parser.prototype.onFrameBody = function onFrameBody(header, buffer, callback) {
...
+  if (this._lastHeaderBlock) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: `this._lastHeaderBlock !== null`

File: lib/spdy-transport/protocol/http2/parser.js [hunk 25]
```diff
@@ -345,7 +364,7 @@ Parser.prototype.onRSTFrame = function onRSTFrame(header, body, callback) {
     return callback(this.error(constants.error.PROTOCOL_ERROR,
                                'Invalid stream id for RST_STREAM'));
   }
-
+
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Unnecessary change, will not pass lint (most likely)

File: lib/spdy-transport/protocol/http2/parser.js [hunk 26]
```diff
@@ -263,6 +277,11 @@ Parser.prototype.queueHeaderBlock = function queueHeaderBlock(header,
         return callback(self.error(constants.error.PROTOCOL_ERROR,
                                    'Header list is too large'));
       }
+      
+      if (header.name.toLowerCase() !== header.name) {
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Gosh, this is potentially a serious performance problem. Any way it could be done simpler?

Perhaps, using `/[A-Z]/.test` should be fine here? Could there be any non-ascii characters?

File: lib/spdy-transport/protocol/http2/parser.js [hunk 27]
```diff
@@ -353,6 +372,28 @@ Parser.prototype.onRSTFrame = function onRSTFrame(header, body, callback) {
   });
 };
 
+Parser.prototype._validateSettings = function _validateSettings(settings) {
+  if ((settings['enable_push'] !== undefined) &&
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: I think there are lots of superfluous parens in this function.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 28]
```diff
@@ -395,7 +441,6 @@ Parser.prototype.onSettingsFrame = function onSettingsFrame(header,
 Parser.prototype.onPushPromiseFrame = function onPushPromiseFrame(header,
                                                                   body,
                                                                   callback) {
-
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Unnecessary change.

File: lib/spdy-transport/stream.js [hunk 29]
```diff
@@ -271,6 +271,9 @@ Stream.prototype._handleWindowUpdate = function _handleWindowUpdate(frame) {
   var state = this._spdyState;
 
   state.window.send.update(frame.delta);
+  if (state.window.send.current > 2147483647) {
```
Comment (indutny) [inline lib/spdy-transport/stream.js]: Considering that this is duplicated, I wonder if it could be moved to `window.js` file? Perhaps some new event may be used to indicate this kind of errors.

I would also pass this constant to the Window on initialization as a part of the options, and would make it inherit it on clone.

**Comment by indutny** [comment]
This is looking great! Fantastic job. Left a series of comments, though.

File: lib/spdy-transport/connection.js (hunk 14 above)
Comment (djphoenix) [inline lib/spdy-transport/connection.js]: ACK

File: lib/spdy-transport/connection.js (hunk 15 above)
Comment (djphoenix) [inline lib/spdy-transport/connection.js]: Nope, there is a case in h2spec that must be handled different with "non-exist" and "closed" streams. Can you propose another solution?

File: lib/spdy-transport/protocol/http2/parser.js (hunk 26 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: ACK

File: lib/spdy-transport/stream.js (hunk 29 above)
Comment (djphoenix) [inline lib/spdy-transport/stream.js]: ACK

File: lib/spdy-transport/connection.js (hunk 18 above)
Comment (djphoenix) [inline lib/spdy-transport/connection.js]: https://github.com/indutny/spdy-transport/pull/23#discussion_r49414492

File: lib/spdy-transport/connection.js (hunk 15 above)
Comment (indutny) [inline lib/spdy-transport/connection.js]: Should it just use last stream id?

Comment (djphoenix) [inline lib/spdy-transport/connection.js]: ACK

File: lib/spdy-transport/protocol/http2/constants.js (hunk 21 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/constants.js]: ACK, done

File: lib/spdy-transport/protocol/http2/parser.js (hunk 23 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: It seems another frames is valid in-between of HEADERS.
h2spec 5.5. Extending HTTP/2 just checks:
    ✓ Sends an unknown extension frame
    ✓ Sends an unknown extension frame in the middle of a header block
So both checks are passed and none of "in the middle" checks in list.

File: lib/spdy-transport/protocol/http2/parser.js (hunk 24 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: ACK

**Comment by djphoenix** [comment]
Take a look again, all seems as fixed.

File: lib/spdy-transport/connection.js [hunk 30]
```diff
@@ -185,6 +197,13 @@ Connection.prototype._init = function init() {
     self.setTimeout(0);
   });
 
+  state.window.recv.on('overflow', function() {
```
Comment (indutny) [inline lib/spdy-transport/connection.js]: Sorry for nit-picking, but I think that this function may be put right above these two `on()` calls, and reused between them.

File: lib/spdy-transport/connection.js [hunk 31]
```diff
@@ -269,13 +288,40 @@ Connection.prototype._handleFrame = function _handleFrame(frame) {
...
+    return;
+  }
+
   if (!stream && frame.id !== undefined) {
     // Load created one
     stream = state.stream.map[frame.id];
 
-    // RST or PRIORITY maybe out of sync
-    if (!stream && (frame.type === 'RST' || frame.type === 'PRIORITY'))
+    if (!stream && state.isServer && state.version === 4 && frame.id !== 0 &&
```
Comment (indutny) [inline lib/spdy-transport/connection.js]: Please add newlines after each `&&` here.

File: lib/spdy-transport/protocol/http2/parser.js [hunk 32]
```diff
@@ -154,9 +154,18 @@ Parser.prototype.onFrameBody = function onFrameBody(header, buffer, callback) {
...
 };
 
+Parser.prototype.onUnknownFrame = function onUnknownFrame(header, buffer, callback) {
+  if (this._lastHeaderBlock !== null) {
+    callback(this.error(constants.error.PROTOCOL_ERROR,
+                        'Received unknown frame in the middle of a header block'));
+    return;
+  }
+  callback(null, { type: 'unknown: ' + header.type });
+}
```
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Missing semicolon

File: lib/spdy-transport/protocol/http2/parser.js (hunk 23 above)
Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: I think this is just odd `h2spec` thing.

This is how it sounds in spec, actually:

```
A CONTINUATION frame MUST be preceded by a HEADERS, PUSH_PROMISE or CONTINUATION frame without the END_HEADERS flag set. A recipient that observes violation of this rule MUST respond with a connection error (Section 5.4.1) of type PROTOCOL_ERROR.
```

File: lib/spdy-transport/stream.js [hunk 33]
```diff
@@ -64,6 +64,14 @@ function Stream(connection, options) {
   this.on('finish', this._onFinish);
   this.on('end', this._onEnd);
 
+  var self = this;
+  state.window.recv.on('overflow', function() {
+    self._onWindowOverflow();
```
Comment (indutny) [inline lib/spdy-transport/stream.js]: Same here, function could be reused.

**Comment by indutny** [comment]
@djphoenix man this looks so lovely! Thank you very much for all your efforts and bearing with my nit-picking. Few very last comments, and it should be good to go!

There seems to be one test failure on travis, though:

```
1) Transport/Connection http2 (v4) should handle SETTINGS:
     Uncaught SETTINGS_MAX_FRAME_SIZE is OOB
```

May I ask you to take a look at it?

File: lib/spdy-transport/connection.js (hunk 30 above)
Comment (djphoenix) [inline lib/spdy-transport/connection.js]: ACK

File: lib/spdy-transport/connection.js (hunk 31 above)
Comment (djphoenix) [inline lib/spdy-transport/connection.js]: ACK

File: lib/spdy-transport/stream.js (hunk 33 above)
Comment (djphoenix) [inline lib/spdy-transport/stream.js]: ACK

**Comment by djphoenix** [comment]
This is fixed in #24. All tests fixes is noted with RFC link/section. So, about "SETTINGS_MAX_FRAME_SIZE is OOB" (https://github.com/djphoenix/spdy-transport/commit/f29ea57f95b8162c440bd004587a03a3c73174eb):
https://tools.ietf.org/html/rfc7540 6.5.2
```
SETTINGS_MAX_FRAME_SIZE (0x5):  Indicates the size of the largest
      frame payload that the sender is willing to receive, in octets.

      The initial value is 2^14 (16,384) octets.  The value advertised
      by an endpoint MUST be between this initial value and the maximum
      allowed frame size (2^24-1 or 16,777,215 octets), inclusive.
      Values outside this range MUST be treated as a connection error
      (Section 5.4.1) of type PROTOCOL_ERROR.
```

File: lib/spdy-transport/protocol/http2/parser.js (hunk 23 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: I think this is another case, covered by h2spec `6.10 CONTINUATION` (handled in another place):
```
    ✓ Sends a CONTINUATION frame after the frame other than HEADERS, PUSH_PROMISE or CONTINUATION
```
This one covers just `5.5. Extending HTTP/2`

Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Hm... I just think that it would be logical to fix both at once, instead of doing two fixes, right?

**Comment by indutny** [comment]
One last nit, otherwise LGTM

File: lib/spdy-transport/protocol/http2/parser.js (hunk 23 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: Hm... I think no because processing of valid (known) frames should be in their own handlers. So this is handler for just invalid/unknown frames. I think this is right, nope?

Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: I don't know, do you want to handle this in every frame? It looks like it is going to be tons of duplicated code across the frames.

What's the reason to keep it separate?

Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: Looks like it is handled at this point, because h2spec is passed this check. So no duplicated code right here.

Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Ah, looks like we was talking about different things!

So what your patch fixes is:

```
HEADERS
CONTINUATION
UNKNOWN
CONTINUATION
```

Your patch makes this sequence fail. What I would actually like to see it doing is making this fail too:

```
HEADERS
CONTINUATION
WHATEVER NON-HEADERS and NON-CONTINUATION frame
CONTINUATION
```

It doesn't seem to be important that it is not currently checked by h2spec, but if this PR will fix the latter issue - the first issue will be fixed automatically, and we will be more spec compliant!

Does this sound reasonable?

Sorry for long discussion on this, I just want to figure out the best way to do it!

**Comment by indutny** [comment]
Overall Pull Request looks fantastic! Clean and beautiful, thank you for your work on this, I appreciate it! Let's figure out that lengthy last nit, and I will land this! :+1:

File: lib/spdy-transport/protocol/http2/parser.js (hunk 23 above)
Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: There is another thing that fixed by this:
```
HEADERS (without (END_HEADERS/FIN)? flag)
UNKNOWN (fail here)
```
There is no `CONTINUATION`-related frames, I think.
Also, unknown-frames are valid to process anywhere else (just ignore it).

About `CONTINUATION` - `queueHeaderBlock` seems to check that you talk about.

Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Heh, the last `CONTINUATION` frames were not necessary in my example. Sorry! The idea is that it should kill connection right on reception of the non-CONTINUATION and non-HEADERS frame. Right now you do it for UNKNOWN frames, but there is no reason to not do it for all non-HEADERS/non-CONTINUATION frames. Is this clear?

Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: I think there is valid to receive frames other than CONTINUATION. Only for CONTINUATION frame we need to check what was received previously. Correct me if not.

Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: It is not valid as far as I far as I see it, because of this:

```
A CONTINUATION frame MUST be preceded by a HEADERS, PUSH_PROMISE or
CONTINUATION frame without the END_HEADERS flag set. A recipient that observes
violation of this rule MUST respond with a connection error (Section 5.4.1) of type
PROTOCOL_ERROR.
```

Basically, CONTINUATION frames are required after HEADERS without END_HEADERS, and no other frames are allowed before the CONTINUATION. So I guess it should be safe to disallow all frames except GOAWAY. Do you agree?

Comment (djphoenix) [inline lib/spdy-transport/protocol/http2/parser.js]: What about RST/SETTINGS/PING (and possible more)? I think they may appear at this point.

Comment (indutny) [inline lib/spdy-transport/protocol/http2/parser.js]: Argh, anyway. Let's do it the way you do it :) I don't care that much about it. We can always figure it out later.

**Comment by indutny** [comment]
Landed and released, thank you so much! You are awesome.
